src/string_enums/enum_base.hpp

```diff
@@ -0,0 +1,65 @@
+/*
+	Copyright (C) 2008 - 2021
+	Part of the Battle for Wesnoth Project https://www.wesnoth.org/
+
+	This program is free software; you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation; either version 2 of the License, or
+	(at your option) any later version.
+	This program is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY.
+
+	See the COPYING file for more details.
+*/
+
+#pragma once
+
+#include <optional>
+#include <string>
+
+namespace string_enums
+{
+/**
+ * The base template for associating string values with enum values.
+ * Implementing classes should not set custom int values for their enum.
+ * The number of enum values must match the number of elements in the @a values array.
```

Comment:
Couldn't this be enforced with a static assert? Something like…

```cpp
static_assert(T::values.size() == T::type::ENUM_MAX);
```

Although that does require adding a redundant ENUM_MAX entry to every enum, so maybe that's unacceptable…

Reply:
I had thought about something similar, but instead of doing an assert, define the array as:
`static constexpr std::array<const char*, static_cast<int>(type::ENUM_MAX)> values{ none, human, ai, reserved };`

I'd been trying to find it there was some built-in way in C++ to find the number of elements in an enum, but it doesn't look like such a feature exists.

Reply:
Yeah, the only way that I know of is adding an extra value to the enum like I mentioned.

Reply:
> I'd been trying to find it there was some built-in way in C++ to find the number of elements in an enum, but it doesn't look like such a feature exists.

Looks like it may eventually be provided by the Reflection TS, but no compiler seems to support that yet.

Reply:
Added.